infrastructure/src/main/java/org/corfudb/infrastructure/NettyRequestRouter.java

```diff
@@ -0,0 +1,164 @@
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Request;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Response;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Header;
+import org.corfudb.runtime.view.Layout;
+
+import java.io.IOException;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+@Slf4j
+@ChannelHandler.Sharable
+public class NettyRequestRouter extends ChannelInboundHandlerAdapter implements IRequestRouter {
+
+    /**
+     * This map stores the mapping from message types to server handler.
+     */
+    private final Map<MessageType, AbstractServer> requestTypeHandlerMap;
+
+    /**
+     * This node's server context.
+     */
+    private final ServerContext serverContext;
+
+    /**
+     * The epoch of this router. This is managed by the base server implementation.
+     */
+    @Getter
+    @Setter
+    volatile long serverEpoch;
+
+    /** The {@link AbstractServer}s this {@link NettyRequestRouter} routes messages for. */
+    private final ImmutableList<AbstractServer> servers;
+
+    /**
+     * Construct a new NettyRequestRouter.
+     * @param servers A list of {@link AbstractServer}s this router will route requests for.
+     * @param serverContext The server context.
+     */
+    public NettyRequestRouter(ImmutableList<AbstractServer> servers, ServerContext serverContext) {
+        this.serverContext = serverContext;
+        serverEpoch = serverContext.getServerEpoch();
+        this.servers = servers;
+        requestTypeHandlerMap = new EnumMap<>(MessageType.class);
+
+        servers.forEach(server -> {
+            Set<MessageType> handledTypes = server.getHandlerMethods().getHandledTypes();
+            handledTypes.forEach(handledType -> requestTypeHandlerMap.put(handledType, server));
+        });
+    }
+
+    @Override
+    public Optional<Layout> getCurrentLayout() {
+        return Optional.ofNullable(serverContext.getCurrentLayout());
+    }
+
+    /**
+     * Send a response message through this router.
+     * @param response The response message to send.
+     * @param ctx The context of the channel handler.
+     */
+    public void sendResponse(Response response, ChannelHandlerContext ctx) {
+        ByteBuf outBuf = PooledByteBufAllocator.DEFAULT.buffer();
+        ByteBufOutputStream responseOutputStream = new ByteBufOutputStream(outBuf);
+
+        try {
+            responseOutputStream.writeByte(API.PROTO_CORFU_MSG_MARK);
+            response.writeTo(responseOutputStream);
+            ctx.writeAndFlush(outBuf, ctx.voidPromise());
+        } catch(IOException e) {
+            log.warn("sendResponse[{}]: Exception occurred when sending response {}, caused by {}",
+                    response.getHeader().getRequestId(), response.getHeader(), e.getCause(), e);
+        } finally {
+            IOUtils.closeQuietly(responseOutputStream);
+        }
+    }
```

Comment:
I would move this functionality into the `AbstractServer`. `ChannelInboundHandlerAdapter` should not concern itself with message serialization.